db/inlineskiplist.h

```diff
@@ -558,36 +681,135 @@ void InlineSkipList<Comparator>::Insert(const char* key) {
+//     (prev[i-1] and prev[i]->Next(i) could be the same node.)
+// In addition prev_height keeps the height of prev[i].
+//
+// When inserting a new key, we look for the lowest level L where
+// prev[L] < key < prev[L-1]. Let
+//    M = max(prev_height[i]..prev_height[num_levels-1])
+// For each level between in [L, M), the previous node of
+// the new key must be one of prev[i]. For levels below L and above M
+// we do normal skip-list search if needed.
+//
+// The optimization is suitable for stream of keys where new inserts are next
+// to or close to the largest key ever inserted, e.g. sequential inserts.
+template <class Comparator>
+void InlineSkipList<Comparator>::InsertWithHint(const char* key,
+                                                InsertHint** hint_ptr) {
+  int height = 0;
+  Node* x = GetNodeForInsert(key, &height);
+
+  // InsertWithHint() is not compatible with prev_ optimization used by
+  // Insert().
+  MaybeInvalidatePrev(height);
+
+  assert(hint_ptr != nullptr);
+  InsertHint* hint = *hint_ptr;
+  if (hint == nullptr) {
+    // AllocateInsertHint will initialize hint with num_levels = 0 and
+    // prev[i] = head_ for all i.
+    hint = AllocateInsertHint();
+    *hint_ptr = hint;
+  }
+
+  // Look for the first level i < num_levels with prev[i] < key.
+  int level = 0;
+  for (; level < hint->num_levels; level++) {
+    if (KeyIsAfterNode(key, hint->prev[level])) {
+      assert(!KeyIsAfterNode(key, hint->prev[level]->Next(level)));
+      break;
+    }
+  }
+  Node* tmp_prev[kMaxPossibleHeight];
+  if (level >= hint->num_levels) {
+    // The hint is not useful in this case. Fallback to full search.
+    FindLessThan(key, tmp_prev);
+    for (int i = 0; i < height; i++) {
+      assert(tmp_prev[i] == head_ || KeyIsAfterNode(key, tmp_prev[i]));
+      assert(!KeyIsAfterNode(key, tmp_prev[i]->Next(i)));
+      x->InsertAfter(tmp_prev[i], i);
+    }
+  } else {
+    // Search on levels below "level", using prev[level] as root.
+    if (level > 0) {
+      FindLessThan(key, tmp_prev, hint->prev[level], level, 0);
+      for (int i = 0; i < level && i < height; i++) {
+        assert(tmp_prev[i] == head_ || KeyIsAfterNode(key, tmp_prev[i]));
+        assert(!KeyIsAfterNode(key, tmp_prev[i]->Next(i)));
+        x->InsertAfter(tmp_prev[i], i);
+      }
+    }
+    // The current level where the new node is to insert into skip-list.
+    int current_level = level;
+    for (int i = level; i < hint->num_levels; i++) {
+      while (current_level < height && current_level < hint->prev_height[i]) {
+        // In this case, prev[i] is the previous node of key on current_level,
+        // since:
+        //   * prev[i] < key;
+        //   * no other nodes less than prev[level-1] has height greater than
+        //     current_level, and prev[level-1] > key.
+        assert(KeyIsAfterNode(key, hint->prev[i]));
+        assert(!KeyIsAfterNode(key, hint->prev[i]->Next(current_level)));
```

Comment:
I still feel we should turn those asserts to actual check, just to be safe. If the check fails, simply fall back to normal Insert().